sympy/matrices/eigen.py

```diff
@@ -1357,6 +1368,144 @@ def pick_vec(small_basis, big_basis):
 
     return restore_floats2(basis_mat, jordan_mat)
 
+def jordan_form_rational_matrix(M, calc_transform):
+    """
+    Return (P, J), where J is the Jordan form and P satisfies M = P J P^{-1},
+    if the matrix is rational. If calc_transform is False, return only J.
+    """
+    from itertools import chain
+
+    def char_mat(lam, mul):
+        # Build (A - lambda * I)^mul over the algebraic field of lam
+        K = QQ.algebraic_field(lam)
+        dM = M.to_DM()
+        return (dM.convert_to(K) - dM.eye(M.shape, K) * K.unit).pow(mul)
+
+    def nullspace_matrix(fac, multiplicity):
+        roots = factor_to_roots[fac]
+        base_root = roots[0]
+        a = AlgebraicNumber(base_root, alias='a')
+        base_nullspace = char_mat(a, multiplicity).nullspace().to_Matrix().T
+        all_vectors = [base_nullspace.subs(a, r)  for r in roots]
+        return all_vectors
+
+    # helper functions
+    def nullity_chain(val, algebraic_multiplicity):
+        """Calculate the sequence  [0, nullity(E), nullity(E**2), ...]
+        until it is constant where ``E = M - val*I``"""
+
+        ret     = [0]
+        cols = M.cols
+        mat_char = char_mat((factor_to_roots[val])[0], 1)
+        mat_pow = mat_char
+        nullity = cols -  mat_char.rank()
+        i       = 2
+
+        while nullity != ret[-1]:
+            ret.append(nullity)
+
+            if nullity == algebraic_multiplicity:
+                break
+            mat_pow = mat_pow * mat_char
+            nullity = cols - mat_pow.rank()
+            i       += 1
+            # Due to issues like #7146 and #15872, SymPy sometimes
+            # gives the wrong rank. In this case, raise an error
+            # instead of returning an incorrect matrix
+            if nullity < ret[-1] or nullity > algebraic_multiplicity:
+                raise MatrixError(
+                    "SymPy had encountered an inconsistent "
+                    "result while computing Jordan block: "
+                    "{}".format(M))
```

Comment:
It is reasonable to check this as a sanity check but I don't think that the comment here about wrongly computed ranks applies when we compute over an algebraic field.

I'm not sure that we really need to compute the rank though. I thought that maybe it is possible to just check the dimension of the nullspaces.